File: src/main/ATOMElement.java
package main;

import java.io.File;
import java.util.Scanner;

public abstract class ATOMElement {

    public ATOMScope parent;

    public static ATOMElement fromTemplate(ATOMTemplate template) {
        char c = template.currentValidChar();
        for (ATOMOperation op : ATOMOperation.operations) {
            for (String opStr : op.commands) {
                if (c==opStr.charAt(0) && template.source.startsWith(opStr, template.currentIndex)) {
                    template.currentIndex += opStr.length();
                    return op;
                }
            }
        }
        if (c=='(') {
            return new ATOMScope(template, ATOMScopeType.PARENTHESIS);
        }
        if (c=='[') {
            return new ATOMScope(template, ATOMScopeType.ARRAY);
        }
        if (c=='{') {
            return new ATOMScope(template, ATOMScopeType.OBJECT);
        }
        // BUNDLER: BEGIN IGNORE
        if (c=='@') {
            int startIndex = template.currentIndex;
            int endIndex = template.currentIndex+1;
            while (isFileChar(template.source.charAt(endIndex))) {
                endIndex++;
                if (endIndex == template.source.length()) {
                    break;
                }
            }
            template.currentIndex = endIndex;
            try {
                File f = new File(template.source.substring(startIndex + 1, endIndex));
                if (f.getName().indexOf(".atom")>0) {
                    return ATOMRuntime.processFile(f);
                } else {
                    String fileContent = "";
                    Scanner in = new Scanner(f);
                    while (in.hasNext()) {
                        fileContent += in.nextLine()+'\n';
                    }
                    return new ATOMValue(fileContent);
                }
            } catch (Exception e) {
                return ATOMValue.NULL();
            }
        }
        // BUNDLER: END IGNORE
        if (c=='"') {
            int startIndex = template.currentIndex;
            int endIndex = template.currentIndex+1;
            while (template.source.charAt(endIndex)!='"') {
                if (template.source.charAt(endIndex)=='\\' && template.source.charAt(endIndex+1)=='"') {
                    endIndex+=2;
                } else {
                    endIndex++;
                }
            }
            template.currentIndex = endIndex+1;
            return new ATOMValue(template.source.substring(startIndex+1, endIndex)
                    .replaceAll("\\\\\"", "\"")
                    .replaceAll("\\\\n", "\n"));
        }
        if (c=='$') {
            int startIndex = template.currentIndex+1;
            int endIndex = template.currentIndex+1;
            boolean isString = false;
            c = template.source.charAt(endIndex);
            while ((c>='0' && c<='9') || isAlphabet(c)) {
                if (isAlphabet(c)) {
                    isString = true;
                }
                endIndex ++;
                c = template.source.charAt(endIndex);
            }
            template.currentIndex = endIndex;
            if (isString) {
                return new ATOMValueDynamic(template.source.substring(startIndex, endIndex), true);
            }
            return new ATOMValueDynamic(Integer.parseInt(template.source.substring(startIndex, endIndex)));
        }
        if (isAlphabet(c)) {
            int startIndex = template.currentIndex;
            int endIndex = template.currentIndex;
            while (isAlphabet(template.source.charAt(endIndex))) {
                endIndex++;
            }
            template.currentIndex = endIndex;
            return new ATOMValueDynamic(template.source.substring(startIndex, endIndex));
        }
        if (c>='0' && c<='9') {
            int startIndex = template.currentIndex;
            int endIndex = template.currentIndex;
            while (template.source.charAt(endIndex)>='0' && template.source.charAt(endIndex)<='9') {
                endIndex++;
            }
            template.currentIndex = endIndex;
            return new ATOMValue(Integer.parseInt(template.source.substring(startIndex, endIndex)));
        }
        throw new RuntimeException(c+"");
    }

    // Eval is lazy
    // They usually will return the same value, except when called on a function
    // Eval will return the function itself
    // Compute will return the result of the function
    public abstract ATOMValue eval();
    public abstract ATOMValue compute();


    private static boolean isAlphabet(char c) {
        return c >= 'A' && c <= 'Z' || c>='a' && c<='z';
    }

    private static boolean isFileChar(char c) {
        return c >= 'A' && c <= 'Z' || c>='a' && c<='z' || c>='0' && c<='9' || c=='_' || c=='/' || c=='.';
    }

}

File: web_interpreter/atom/ATOMElement.js
import ATOMOperation from './ATOMOperation.js';
import ATOMScope from './ATOMScope.js';
import ATOMValue from './ATOMValue.js';
import ATOMValueDynamic from './ATOMValueDynamic.js';

export default class ATOMElement {

  constructor() {
    this.parent = null;
  }

  static fromTemplate(atomTemplate) {
    let c = atomTemplate.currentValidChar();
    for (let op of ATOMOperation.operations) {
      for (let opStr of op.commands) {
        if (c==opStr[0] && atomTemplate.source.substring(atomTemplate.currentIndex, atomTemplate.currentIndex+opStr.length)==opStr) {
          atomTemplate.currentIndex += opStr.length;
          return op;
        }
      }
    }
    if (c=='(') {
      return new ATOMScope(atomTemplate, 'PARENTHESIS');
    }
    if (c=='[') {
      return new ATOMScope(atomTemplate, 'ARRAY');
    }
    if (c=='{') {
      return new ATOMScope(atomTemplate, 'OBJECT');
    }
    if (c=='"') {
      let startIndex = atomTemplate.currentIndex;
      let endIndex = atomTemplate.currentIndex+1;
      while (atomTemplate.source[endIndex]!='"') {
        if (atomTemplate.source[endIndex]=='\\' && atomTemplate.source[endIndex+1]=='"') {
          endIndex+=2;
        } else {
          endIndex++;
        }
      }
      atomTemplate.currentIndex = endIndex+1;
      return new ATOMValue(atomTemplate.source.substring(startIndex+1, endIndex).replace(/\\"/g, '"'));
    }
    if (c=='$') {
      let startIndex = atomTemplate.currentIndex+1;
      let endIndex = atomTemplate.currentIndex+1;
      let isString = false;
      c = atomTemplate.source[endIndex];
      while ((c>='0' && c<='9') || this.isAlphabet(c)) {
        if (this.isAlphabet(c)) {
          isString = true;
        }
        endIndex ++;
        c = atomTemplate.source[endIndex];
      }
      atomTemplate.currentIndex = endIndex;
      if (isString) {
        return new ATOMValueDynamic(atomTemplate.source.substring(startIndex, endIndex), true);
      }
      return new ATOMValueDynamic(parseInt(atomTemplate.source.substring(startIndex, endIndex)));
    }
    if (this.isAlphabet(c)) {
      let startIndex = atomTemplate.currentIndex;
      let endIndex = atomTemplate.currentIndex;
      while (this.isAlphabet(atomTemplate.source[endIndex])) {
        endIndex++;
      }
      atomTemplate.currentIndex = endIndex;
      return new ATOMValueDynamic(atomTemplate.source.substring(startIndex, endIndex));
    }
    if (c>='0' && c<='9') {
      let startIndex = atomTemplate.currentIndex;
      let endIndex = atomTemplate.currentIndex;
      while (atomTemplate.source[endIndex]>='0' && atomTemplate.source[endIndex]<='9') {
        endIndex++;
      }
      atomTemplate.currentIndex = endIndex;
      return new ATOMValue(parseInt(atomTemplate.source.substring(startIndex, endIndex)));
    }
    console.log('NO MATCH FOUND FOR CHAR CODE')
    console.log((''+c).charCodeAt(0))
    console.log('INDEX IS');
    console.log(atomTemplate.currentIndex);
    throw c;
  }

  static isAlphabet(c) {
    return c >= 'A' && c <= 'Z' || c>='a' && c<='z';
  }

}